### PR TITLE
[Misc] DeepGemmExperts : Avoid JIT generation in the hot-path

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
## Purpose
DeepGemm JITs the grouped-gemm kernels. This JIT generation depends on the Gemm input tensor shapes. We want to avoid generating JITs during actual model inference. To this effect, during model profile run, we invoke the DeepGemm grouped_gemm function with all possible input shapes for the model so all JIT generation is complete.

## Test Plan

vllm_serve-command : `VLLM_ALL2ALL_BACKEND="deepep_high_throughput" VLLM_USE_DEEP_GEMM=1   canhazgpu run -g2 -- vllm serve Qwen/Qwen3-30B-A3B-FP8  --trust-remote-code --enable-expert-parallel --data-parallel-size 2 --port 9010 --no-enable-prefix-caching`
lm_eval-command : `lm_eval --model local-completions --tasks gsm8k --model_args model=Qwen/Qwen3-30B-A3B-FP8,base_url=http://127.0.0.1:9010/v1/completions,num_concurrent=30,max_retries=3 --limit 100 `

## Test Result
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.85|±  |0.0359|
|     |       |strict-match    |     5|exact_match|↑  | 0.94|±  |0.0239|
```

## Test JIT Generation
DeepGemm stores the JITs at `~/.deep_gemm/cache`.
- Clear `~/.deep_gemm/cache`
- Execute `vllm_serve-command`
- After engine startup, I verified that the JITs are populated at `~/.deep_gemm/cache`
- Execute `lm_eval-command`.
- After the `lm_eval` run command finished I verified that there were no new JITs created.

### CODE REVIEW

**Comment by varun-sundar-rabindranath** [comment]
@tlrmchlsmth @bnellnm @yewentao256  PTAL! Thanks ! 

One threat is the extra memory allocation. We need to test this PR with multi-node `Deepseek-R1` before landing.

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py [hunk 1]
```diff
@@ -82,6 +82,61 @@ def _valid_deep_gemm(hidden_states: torch.Tensor, w1: torch.Tensor,
...
+                              num_experts,
+                              deep_gemm_block_shape()[0],
+                              expert_tokens_meta=None)
+    expert_ids = torch.zeros((MAX_M, ), device=device, dtype=torch.int32)
+
+    def _warmup(w: torch.Tensor, w_scale: torch.Tensor):
+
+        _, n, k = w.size()
+        a1q = torch.empty((MAX_M, k), device=device).to(torch.float8_e4m3fn)
+        a1q_scales = torch.empty((MAX_M, k // 128),
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: should this 128 come from `deep_gemm_block_shape`?

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py [hunk 2]
```diff
@@ -156,6 +211,19 @@ def apply(
...
         assert a1q_scale is not None
+        assert w1_scale is not None
+        assert w2_scale is not None
+
+        # DeepGemm JITs the grouped-gemm kernels. We don't want the JIT'ing
+        # to happen during actual model-inference. The `warmup_deepgemm_kernels`
+        # function is a `run_once` decorated that executes during the model
+        # profile run. This warmup is should create all the required JITs for
+        # this model.
+        warmup_deepgemm_kernels(w1,
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: The placement of this warmup bugs me a bit.  I feel like it should be done in layer.py or something instead (I realize that would be more difficult to implement).  Although, maybe after I land my refactoring PR we could move this to ` __init__` since I plan on moving a bunch of the existing parameters into data members.

Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: actually, I realize that the naming is a bit misleading. This function only runs warmup for `m_grouped_fp8_gemm_nt_contiguous` function and not all deepgemm kernel types. let me rename. 
Also, the warmups are very much dependent on how the Gemms are used (for example, it depends on  `compute_aligned_M` and weight shapes that are very specific to this layer's implementation) - in that sense it is good that these warmups stay close to where they are used ?

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: > actually, I realize that the naming is a bit misleading. This function only runs warmup for `m_grouped_fp8_gemm_nt_contiguous` function and not all deepgemm kernel types. let me rename. Also, the warmups are very much dependent on how the Gemms are used (for example, it depends on `compute_aligned_M` and weight shapes that are very specific to this layer's implementation) - in that sense it is good that these warmups stay close to where they are used ?

Yeah, I realize that.  It still seems weird to have it right in the middle of `apply` tho.   Once my refactoring PR lands I think this could probably go in `__init__` instead.

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py [hunk 3]
```diff
@@ -82,6 +82,61 @@ def _valid_deep_gemm(hidden_states: torch.Tensor, w1: torch.Tensor,
...
+    device = w1.device
+
+    # This is the maximum GroupedGemm M size that we expect to run
+    # the grouped_gemm with.
+    MAX_M = compute_aligned_M(env.VLLM_FUSED_MOE_CHUNK_SIZE,
+                              num_topk,
+                              num_experts,
+                              deep_gemm_block_shape()[0],
+                              expert_tokens_meta=None)
+    expert_ids = torch.zeros((MAX_M, ), device=device, dtype=torch.int32)
```
[the remark this answers is not in the document]
Comment (tlrmchlsmth) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: Might be good to do what the bot says, out of paranoia

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py [hunk 4]
```diff
@@ -82,6 +82,62 @@ def _valid_deep_gemm(hidden_states: torch.Tensor, w1: torch.Tensor,
...
+        num_tokens = MAX_M
+        while num_tokens > 0:
+            m_grouped_fp8_gemm_nt_contiguous(
+                (a1q[:num_tokens], a1q_scales[:num_tokens]), (w, w_scale),
+                out[:num_tokens], expert_ids[:num_tokens])
+            pbar.update(1)
+            num_tokens = num_tokens - block_m
+
+    _warmup(w1, w1_scale)
+    _warmup(w2, w2_scale)
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: Do you know how long this process takes @varun-sundar-rabindranath?

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 2 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: I see. Yeah, I agree. Thanks  🙌

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 4 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: It takes around 30-45 seconds when the cache is empty.

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: I considered the suggestion and thought there wouldn't be hashing (for jit'ing)  on the contents of the GPU tensor. 

But I agree there is no loss in doing it anyways. I'll fix it.

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 4 above)
Comment (tlrmchlsmth) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: Especially once the jitting is already done and the artifacts saved on disk (?), we might want to provide a way for users to skip this process to support faster model loading

**Review by tlrmchlsmth — COMMENTED** [review]
Is it possible to change the directory that the jitted kernels get saved to?

Instead of `$HOME/.deep_gemm/cache`, we should stick them in `~/.cache/vllm/deep_gemm/cache` so that there's one consistent place to keep track of all of vLLM's cached files

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py [hunk 5]
```diff
@@ -82,6 +82,69 @@ def _valid_deep_gemm(hidden_states: torch.Tensor, w1: torch.Tensor,
...
+    DeepGemm JITs the grouped-gemm kernels. The JIT'ing happens based on the
+    input tensor shapes. In this function, we construct all possible input
+    tensor shapes so all the kernels are JIT'ed and cached.
+    Note that this warmup is expected to happen during the model profile
+    call and not during actual model inference.
+    """
+
+    from tqdm import tqdm
+
+    import vllm.envs as env
```
Comment (yewentao256) [inline vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: Is there any reason we prefer put import inside the function?

**Review by yewentao256 — APPROVED** [review]
Looks good to me, thanks for the work!

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 4 above)
Comment (yewentao256) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: Since it is just 30-45 seconds for the first time, so perhaps no need to add a control for user?

Comment (yewentao256) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: +1, curious about how long it would take when it is done once

Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: when it is done once (i.e. per model) it takes less than 3 seconds.

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 5 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: I am not sure if this is the standard. I have been doing it to make imports lazy. also it adds the context that the imports are just used within the function.

Comment (yewentao256) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: I think usually if there is no circular import, we prefer put high level import in top for easier maintainence, and little performance improvement

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 4 above)
Comment (yewentao256) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: Make sense, then I think there is no need to skip when it is done

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 5 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: yeah - I see the argument. I have moved the imports to the top 👍

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py [hunk 6]
```diff
@@ -156,6 +217,20 @@ def apply(
...
+            # DeepGemm JITs the grouped-gemm kernels. We don't want the JIT'ing
+            # to happen during actual model-inference. The
+            # `warmup_deepgemm_kernels` function is a `run_once` decorated
+            # function that executes during the model profile run. This warmup
+            # should create all the required JITs for the current model.
+            warmup_deepgemm_gg_contiguous_kernels(w1,
+                                                  w2,
+                                                  w1_scale,
+                                                  w2_scale,
+                                                  num_topk=topk_ids.size(1))
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/deep_gemm_moe.py:233]: Added an environment flag to skip warmups.

File: vllm/utils/deep_gemm.py [hunk 7]
```diff
@@ -69,6 +70,10 @@ def _lazy_init() -> None:
     if not has_deep_gemm():
         return
 
+    # Set up deep_gemm cache path
+    os.environ['DG_JIT_CACHE_DIR'] = os.path.join(envs.VLLM_CACHE_ROOT,
+                                                  "deep_gemm")
+
```
Comment (varun-sundar-rabindranath) [inline vllm/utils/deep_gemm.py:78]: Set `DG_JIT_CACHE_DIR` so the jit cache is created at `~/.cache/vllm/deep_gemm/` . without this it would be created at `~/.deep_gemm/`

Comment (varun-sundar-rabindranath) [reply vllm/utils/deep_gemm.py:78]: @yewentao256  I looked through the code base and this place (i.e. inside `lazy_init`) appeared to be the best place for this setting. what do you think ?

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 4 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py]: Add an env flag to skip the warmup https://github.com/vllm-project/vllm/pull/21955/files#r2248050738

File: vllm/utils/deep_gemm.py (hunk 7 above)
Comment (varun-sundar-rabindranath) [reply vllm/utils/deep_gemm.py:78]: cc @tlrmchlsmth

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 6 above)
Comment (tlrmchlsmth) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py:233]: Would the name `VLLM_SKIP_DEEP_GEMM_WARMUP` more clearly convey intent?

File: vllm/utils/deep_gemm.py (hunk 7 above)
Comment (tlrmchlsmth) [reply vllm/utils/deep_gemm.py:78]: Here is ok with me - no harm to setting the env unconditionally either

Comment (tlrmchlsmth) [reply vllm/utils/deep_gemm.py:78]: I think we should respect the environment variable if it's already set though

File: vllm/model_executor/layers/fused_moe/deep_gemm_moe.py (hunk 6 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/deep_gemm_moe.py:233]: Agreed and updated 👍

File: vllm/utils/deep_gemm.py (hunk 7 above)
Comment (varun-sundar-rabindranath) [reply vllm/utils/deep_gemm.py:78]: Yes. I missed that scenario. Updated 👍

Comment (yewentao256) [reply vllm/utils/deep_gemm.py:78]: > @yewentao256 I looked through the code base and this place (i.e. inside `lazy_init`) appeared to be the best place for this setting. what do you think ?

Sounds good to me

**Review by tlrmchlsmth — APPROVED** [review]
